feat(entrypoint-0.7): base - all changes in regards to the Entrypoint v6 & v7 support

denniswon opened the pull request:
# Pull Request Checklist

- [ ] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [ ] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [ ] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [ ] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [ ] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

<!-- start pr-codex -->

---

## PR-Codex overview
This PR updates various files across packages to introduce versioning changes related to entry points, smart accounts, and APIs.

### Detailed summary
- Renamed entry point ABI files to include version suffix
- Updated `HexSchema` to use strict hex validation
- Added `API_KEY_STAGING` constant in test constants
- Modified functions to use `Promise<NullAddress | Address>`
- Updated function signatures to include versioning information
- Introduced new error classes for entry point and user operation
- Refactored middleware functions to handle different entry point versions

> The following files were skipped due to too many changes: `packages/core/src/utils/testUtils.ts`, `packages/core/src/errors/account.ts`, `packages/accounts/src/msca/plugin-manager/decorator.ts`, `packages/core/src/account/schema.ts`, `packages/ethers/src/types.ts`, `packages/alchemy/src/middleware/userOperationSimulator.ts`, `packages/accounts/src/light-account/decorator.ts`, `packages/plugingen/src/commands/generate/phases/plugin-actions/management-actions.ts`, `packages/core/src/middleware/types.ts`, `packages/ethers/src/provider-adapter.ts`, `packages/core/src/actions/smartAccount/sendTransactions.ts`, `packages/core/src/middleware/defaults/overridePaymasterData.ts`, `packages/accounts/src/msca/plugin-manager/uninstallPlugin.ts`, `packages/accounts/src/msca/plugins/multisig/actions/proposeUserOperation.ts`, `packages/accounts/src/msca/plugins/multisig/types.ts`, `packages/accounts/src/msca/plugin-manager/installPlugin.ts`, `packages/core/src/actions/smartAccount/sendUserOperation.ts`, `packages/plugingen/src/commands/generate/phases/plugin-actions/read-actions.ts`, `packages/accounts/plugingen/phases/plugin-actions/management-actions.ts`, `packages/accounts/src/light-account/actions/transferOwnership.ts`, `packages/core/src/actions/smartAccount/checkGasSponsorshipEligibility.ts`, `packages/accounts/src/nani-account/account.ts`, `packages/core/src/actions/smartAccount/signUserOperation.ts`, `packages/accounts/src/msca/plugins/multi-owner/plugin.ts`, `packages/accounts/src/msca/plugins/multisig/plugin.ts`, `packages/core/src/index.ts`, `packages/accounts/src/msca/plugins/multisig/utils/splitAggregatedSignature.ts`, `packages/accounts/src/msca/account/multisigAccount.ts`, `packages/core/src/actions/smartAccount/sendTransaction.ts`, `packages/accounts/src/msca/utils.ts`, `packages/core/src/actions/smartAccount/buildUserOperationFromTxs.ts`, `packages/core/src/__tests__/utils.test.ts`, `packages/core/src/account/base.ts`, `packages/core/src/actions/smartAccount/internal/sendUserOperation.ts`, `packages/ethers/src/account-signer.ts`, `packages/core/src/actions/smartAccount/buildUserOperationFromTx.ts`, `packages/core/src/utils/types.ts`, `packages/core/src/actions/smartAccount/internal/runMiddlewareStack.ts`, `packages/core/src/actions/smartAccount/buildUserOperation.ts`, `packages/core/src/utils/index.ts`, `packages/accounts/src/msca/account/multiOwnerAccount.ts`, `packages/core/src/client/decorators/bundlerClient.ts`, `packages/core/src/utils/defaults.ts`, `packages/accounts/plugingen/phases/plugin-actions/index.ts`, `packages/accounts/src/msca/plugins/session-key/plugin.ts`, `packages/core/src/client/decorators/smartAccountClient.ts`, `packages/accounts/src/light-account/account.ts`, `packages/core/src/utils/userop.ts`, `packages/alchemy/src/client/types.ts`, `packages/plugingen/src/commands/generate/phases/plugin-actions/index.ts`, `packages/core/src/entrypoint/0.6.ts`, `packages/accounts/src/msca/plugins/session-key/extension.ts`, `packages/core/src/actions/smartAccount/dropAndReplaceUserOperation.ts`, `packages/core/src/entrypoint/types.ts`, `packages/core/src/actions/smartAccount/types.ts`, `packages/core/src/entrypoint/0.7.ts`, `packages/core/src/entrypoint/index.ts`, `packages/core/e2e-tests/simple-account-staging.e2e.test.ts`, `packages/core/src/account/simple.ts`, `packages/core/src/abis/SimpleAccountAbi_v7.ts`, `packages/core/src/account/smartContractAccount.ts`, `packages/alchemy/src/middleware/gasManager.ts`, `packages/core/src/types.ts`, `packages/core/src/abis/EntryPointAbi_v7.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

<!-- end pr-codex -->

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206854921431396
  - https://app.asana.com/0/0/1206826573496436
  - https://app.asana.com/0/0/1206826573496426
  - https://app.asana.com/0/0/1206924039831384

moldy530 commented on packages/core/src/actions/smartAccount/internal/sendUserOperation.ts:
```diff
@@ -16,10 +17,18 @@ export const _sendUserOperation: <
+    TTransport,
+    TChain,
+    TAccount,
+    TEntryPointVersion
```
why's the client need to be generic on the entrypoint version?

moldy530 replied:
doesn't this result in potential drift between the definition of the entrypoint version here and the version on the account?

moldy530 commented on packages/core/src/actions/smartAccount/types.ts:
```diff
@@ -64,9 +70,11 @@ export type WaitForUserOperationTxParameters = {
+  uoStruct: UserOperationStruct<TEntryPointVersion>;
   batch: BatchUserOperationCallData;
-  overrides: UserOperationOverrides;
+  overrides: UserOperationOverrides<TEntryPointVersion>;
```
what in overrides is entrypoint version dependent?

denniswon replied:
paymaster data is different.

moldy530 commented on packages/core/src/client/bundlerClient.ts:
```diff
@@ -21,35 +22,52 @@ import {
 
-export function createBundlerClient(
-  args: PublicClientConfig & { type?: string }
+export function createBundlerClient<
+  TTransport extends Transport,
+  TEntryPointVersion extends EntryPointVersion = EntryPointVersion
+>(
+  args: PublicClientConfig<TTransport, Chain> & { type?: string } & {
+    entryPointVersion?: TEntryPointVersion;
+  }
```
wait this shouldn't have changed, the overloaded definition was meant to be defined on the defaults as it was before

moldy530 commented on packages/core/src/client/bundlerClient.ts:
```diff
@@ -21,35 +22,52 @@ import {
+  TEntryPointVersion extends EntryPointVersion = EntryPointVersion
+>(
+  args: PublicClientConfig<TTransport, Chain> & { type?: string } & {
+    entryPointVersion?: TEntryPointVersion;
```
where is this used, why's it necessary?

moldy530 commented on packages/core/src/entrypoint/0.6.ts:
why'd we remove this? should we have separate impls for each version? looking down below we handle that by having a switch statement inside of `getUserOperationHash` and `packUo`. That doesn't scale well even with just 2 versions.

Consider how one would leverage an `interface` in OOP. you wouldn't define one `Dog` interfaces that switches internally on the breed, you'd have impls for each breed.

Similarly here, we should have an `Entrypoint` interface, but each version should be defined independently

moldy530 commented on packages/core/src/entrypoint/types.ts:
```diff
@@ -1,9 +1,20 @@
-export type EntryPointDef<EntryPointUserOperation> = {
-  version: string;
+export type EntryPointDef<
+  TEntryPointVersion extends EntryPointVersion | undefined =
+    | EntryPointVersion
+    | undefined
```
when would this ever be undefined?

moldy530 commented on packages/core/src/entrypoint/types.ts:
```diff
@@ -1,9 +1,20 @@
+    | EntryPointVersion
+    | undefined
+> = {
+  version: TEntryPointVersion;
```
I don't think we need to make this generic tbh... this version field in the def can just be `EntrypointVersion` and then impls can be more restrictive in their typing by doing something like `version: 0.6.0 as const` or `version: 0.7.0 as const`

moldy530 commented on packages/core/src/errors/entrypoint.ts:
```diff
@@ -1,15 +1,24 @@
+export class UnexpectedEntryPointVersionError extends BaseError {
+  override name = "UnexpectedEntryPointVersionError";
+
+  constructor(version: any) {
```
shouldn't this be a `string` when would `version` be anything else?

moldy530 commented on packages/core/src/types.ts:
```diff
@@ -48,32 +53,50 @@ export type UserOperationFeeOptions = z.input<
-  nonceKey: bigint;
-}>;
+export type UserOperationOverrides<
+  TEntryPointVersion extends EntryPointVersion = EntryPointVersion
```
👍  this one makes sense to be generic

moldy530 commented on packages/core/src/types.ts:
```diff
@@ -97,6 +120,55 @@ export interface UserOperationRequest {
+  /* anti-replay parameter. nonce of the transaction, returned from the entry point for this address */
+  nonce: Hex;
+  /* account factory, only for new accounts */
+  factory: Address | NullAddress;
```
what's `NullAddress` type?

moldy530 commented on packages/core/src/utils/userop.ts:
```diff
@@ -72,3 +144,183 @@ export function applyUserOpOverrideOrFeeOption(
+  }
+}
+
+export function packAccountGasLimits(
+  data:
+    | Pick<UserOperationRequest_v7, "verificationGasLimit" | "callGasLimit">
+    | Pick<UserOperationRequest_v7, "maxPriorityFeePerGas" | "maxFeePerGas">
+): Hex {
+  return concat(Object.values(data).map((v) => pad(v, { size: 16 })));
+}
+
+export function packPaymasterData({
+  paymaster,
+  paymasterVerificationGasLimit,
+  paymasterPostOpGasLimit,
+  paymasterData,
+}: Pick<
+  UserOperationRequest_v7,
+  | "paymaster"
+  | "paymasterVerificationGasLimit"
+  | "paymasterPostOpGasLimit"
+  | "paymasterData"
+>): Hex {
+  return concat([
+    paymaster,
+    pad(paymasterVerificationGasLimit, { size: 16 }),
+    pad(paymasterPostOpGasLimit, { size: 16 }),
+    paymasterData,
+  ]);
+}
+
+export function unpackAccountGasLimits(accountGasLimits: string): {
+  verificationGasLimit: number;
+  callGasLimit: number;
+} {
+  return {
+    verificationGasLimit: parseInt(accountGasLimits.slice(2, 34), 16),
+    callGasLimit: parseInt(accountGasLimits.slice(34), 16),
+  };
+}
```
yea these are things I'd expect to be defined on specific `EntrypointDef` impls

moldy530 commented on packages/core/src/utils/userop.ts:
```diff
@@ -1,32 +1,104 @@
-export function isValidRequest(
-  request: UserOperationStruct
-): request is UserOperationRequest {
+export function isValidRequest<
```
we talked about not using `zod` in a lot of places... but parsing UserOperationStruct for validity is actually a good usecase.

if the the `EntryPointDef` had a method called `validateUserOperation` we could actually have the `EntryPoint` enforce the types. Then methods like `sendUserOperation` wouldn't necessarily need to be generic on the version and instead would just accept the union type of `UserOperationStructs`. Then the `account.entrypoint` would validate it supports the UO struct passed into it

moldy530 commented on packages/core/src/utils/userop.ts:
```diff
@@ -72,3 +144,183 @@ export function applyUserOpOverrideOrFeeOption(
+
+    case "0.6.0":
+      const hashedInitCode = keccak256(
+        (request as UserOperationRequest<"0.6.0">).initCode
```
going off the above comment. it would omit having to do this `as` since the EP itself parses the UO for validity before parsing it

moldy530 commented on packages/core/src/account/smartContractAccount.ts:
```diff
@@ -179,24 +191,27 @@ export async function toSmartContractAccount<
   const entryPointContract = getContract({
     address: entryPoint.address,
-    abi: EntryPointAbi,
+    abi: entryPoint.version === "0.6.0" ? EntryPointAbi_v6 : EntryPointAbi_v7,
```
maybe we should move the abi into `EntrypointDef` itself so that it handles it

moldy530 commented on packages/core/src/account/smartContractAccount.ts:
```diff
@@ -164,12 +175,13 @@ export const getAccountAddress = async ({
   chain,
   source,
-  entryPoint = getVersion060EntryPoint(chain),
+  entryPoint = getEntryPoint(chain),
```
the current definition of `getEntryPoint` isn't typesafe because the `TEntryPointVersion` doesn't necessarily line up with the version being returned by this. someone could specify `0.6.0` for `TEntryPointVersion` but there's nothing stopping someone from calling this with `getEntryPoint(chain, "0.7.0")`

moldy530 replied:
oh never mind just saw the updated params

moldy530 commented on packages/core/src/actions/smartAccount/signUserOperation.ts:
```diff
@@ -16,11 +17,15 @@ export const signUserOperation: <
   client: Client<TTransport, TChain, TAccount>,
-  args: SignUserOperationParameters<TAccount>
-) => Promise<UserOperationRequest> = async (client, args) => {
+  args: SignUserOperationParameters<TAccount, TEntryPointVersion>
```
I don't think signer user operation should know about the Entrypoint version. if the account has the entrypoint definition attached to it, then shouldn't it be enough to just use `account.entrypoint.blah`?

moldy530 commented on packages/core/src/actions/smartAccount/types.ts:
```diff
@@ -14,46 +15,51 @@ import type {
 > = {
   requests: RpcTransactionRequest[];
-  overrides?: UserOperationOverrides;
+  overrides?: UserOperationOverrides<TEntryPointVersion>;
```
since we already have an account param available to us in these requests, then we can infer the `TEntryPointVersion` and remove the need to define it here. you can use `TAccount["entryPoint"]["version"]` if it's defined or use the `account` param from the `GetAccountParameter<TAccount>` if it's not defined. You can have a helper type to this like `GetEntrypointVersionFromAccount` or something. that keeps the EP responsibility on the account itself rather than having to put it on the client too

moldy530 replied:
this applies to this whole file, I don't think we need to make these methods generic on `EPVersion`

denniswon replied:
but what about bundler client? that one also needs entry point version, but not attached to account. @moldy530

moldy530 replied:
I don't think it does does it? you can make the bundler methods take a union type instead of having to pass in the generic

denniswon replied:
hmm i see, so tldr, client not attached to smart account can be used for either v6 and v7.

moldy530 replied:
if it defaults to the union type already, why not just remove the generic altogether? IIRC the way viem handles the various TX types is a union type without using generics

denniswon replied:
because dealing with union types is not ideal for devs, as you have to do lots of repetitive checking for which entry point version struct/response it is, and you have to cast types so makes the code for devs who use SDK very ugly.

I still think we should use generic for better typescript experience vs using union types.

moldy530 replied:
when using the `bundlerClient` a dev doesn't need to do type checking though. the method just directly supports either format. Looking at how `viem` handles this here https://github.com/wevm/viem/blob/main/src/actions/wallet/sendTransaction.ts#L49 and here https://github.com/wevm/viem/blob/main/src/types/transaction.ts#L231 they also just use a union type. Where they leverage the generic is if the actual request types require something to be generic (which in our case neither of the 2 types require that). 

I would rather we stay in line with how the bundler handles it too which is to use Union types

denniswon commented on packages/core/src/account/schema.ts:
```diff
@@ -24,7 +24,10 @@ export const createBaseSmartAccountParamsSchema = <
       .string()
-      .refine(isHex, "initCode must be a valid hex.")
+      .refine(
+        (x) => isHex(x, { strict: true }),
```
without strict, any string in form `0x{string}` is considered a valid hex. so things like `0xmoldy` considered as Hex 
https://viem.sh/docs/utilities/isHex.html#optionsstrict

moldy530 replied:
ah nice

denniswon commented on packages/core/src/utils/userop.ts:
```diff
@@ -1,32 +1,104 @@
+  // either all exist, or none.
+  return allEqual(
+    isAddress(paymaster),
+    toHex(paymasterData || "0x") !== "0x",
```
paymaster fields (paymaster, paymasterData, paymasterValidationGasLimit, paymasterPostOpGasLimit) either all exist, or none.

https://github.com/ethereum/ercs/blob/master/ERCS/erc-4337.md#rpc-methods-eth-namespace.

moldy530 replied:
hmm yea but the docs do say this too
> All fields MUST be set as hex values. empty bytes block (e.g. empty initCode) MUST be set to "0x"

reading that, I'd imagine that either all the fields are `0x` or all are `Hex`

denniswon replied:
this is what I referenced: https://github.com/eth-infinitism/account-abstraction/blob/8fae866ea908f517f0399edd27057b2cc275576d/test/UserOp.ts#L28 for packing user operation.

So when packing user operation, if it fails to hold that all paymaster-related fields (4 of them) are not null, paymasterAndData for packed uo is encoded as 0x. But per the user operation request sent to the bundler, they are unset.

cc: @dancoombs @dphilipson can you confirm on this logic?

denniswon replied:
```
**Parameters:**
- UserOperation a full user-operation struct. All fields MUST be set as hex values. empty bytes block (e.g. empty initCode) MUST be set to "0x"
- factory and factoryData - either both exist, or none
- paymaster fields (paymaster, paymasterData, paymasterValidationGasLimit, paymasterPostOpGasLimit) either all exist, or none.
- EntryPoint the entrypoint address the request should be sent through. this MUST be one of the entry points returned by the supportedEntryPoints rpc call.
```

denniswon replied:
When it says "either all exist, or none", that means fields are "unset" or set to "0x" when sending user operation?

denniswon replied:
I do get the user op hash returned by the rundler on staging with unexisting fields unset, and in case of failure, the EIP says it MUST return an error result object, with code: -32602 - invalid UserOperation struct/fields. I don't get this error with the current way of omitting unexisting fields, so I assumed this understanding. cc: @moldy530

moldy530 commented on packages/core/src/utils/userop.ts:
```diff
@@ -1,32 +1,104 @@
+>(request: UserOperationStruct<TEntryPointVersion>): boolean {
+  if (!("paymaster" in request)) {
+    const { paymasterAndData } = request as UserOperationStruct_v6;
+    return paymasterAndData != null;
```
for this paymasterAndData must be set as `0x` if it's not being used. Tested this out with @adam-alchemy the other day and we got an error from the RPC

dphilipson commented on packages/core/src/entrypoint/index.ts:
```diff
@@ -0,0 +1,189 @@
+// // 1. left and right hand side version type should match, as well as the param
+
+// // === Error ===
+// const ep1_a: EntryPointDef<"0.6.0"> = getEntryPoint(chain, {
```
You may know this already, but if you're trying to write tests that certain expressions do _not_ typecheck, you can use `// @ts-expect-error` directives ([docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments)).

denniswon replied:
follow up PR moves this to actual unit test

dancoombs commented on packages/core/src/entrypoint/0.6.ts:
```diff
@@ -1,28 +1,115 @@
+export default {
+  version: "0.6.0",
+
+  address: {
```
I really think we should avoid any hardcoding of networks in "core." Is there a way around this?

dphilipson replied:
We could just assume the same address for all networks. The things to keep an eye on:

* If some network uses a different address. I don't know why that would happen, but it's possible.
* We'd need handling for when a network doesn't have an entry point deployed. Potentially we'd have a check somewhere that this address has code on the selected network.

Definitely agree that not hardcoding networks would be great. I think if we didn't need this mapping, we could simplify some types by not needing a `TChain` type parameter on `SupportedEntryPoint` and related types.

moldy530 replied:
I'm ok with hardcoding some defaults, as long as we have some means of overwriting them like we do for account addresses, factories, etc

moldy530 replied:
@denniswon how do we handle chain and address overrides here? what happens if we don't support a chain in the SDK yet, but someone wants to use the EP on a chain our infra or their infra supports?

moldy530 replied:
bump

dancoombs commented on packages/core/src/entrypoint/0.7.ts:
```diff
@@ -0,0 +1,180 @@
+export default {
+  version: "0.7.0",
+
+  address: {
```
Same here, we should avoid hardcoding networks in core.

dphilipson commented on packages/core/src/account/smartContractAccount.ts:
```diff
@@ -14,10 +15,14 @@ import {
+  EntryPointDef,
+  EntryPointDefRegistry,
+  EntryPointVersion,
+} from "../entrypoint/types.js";
```
Never mind, looks like that's just how this codebase is set up.

moldy530 replied:
it's a typescript and ESM thing, you have to include the extension

denniswon commented on packages/core/src/entrypoint/index.ts:
```diff
@@ -0,0 +1,189 @@
+  const { version, addressOverride } = options ?? {
+    version: defaultEntryPointVersion,
+  };
+  const entryPoint = entryPointRegistry[version ?? defaultEntryPointVersion];
```
we do, because options can be just { addressOverride }

dphilipson replied:
Yeah, I mean that you do need the `version ?? defaultEntryPointVersion`, but you don't need the `version: defaultEntryPointVersion` two lines up. If you were to leave that out, then in the null options case `version` would evaluate to `undefined` and then `version ?? defaultEntryPointVersion` would evaluate to `defaultEntryPointVersion` which works out the same as the code as written.

denniswon commented on packages/core/src/entrypoint/types.ts:
```diff
@@ -1,9 +1,124 @@
+> = {
+  version: TEntryPointVersion;
+  address: Record<TChain["id"], Address>;
+  abi: GetContractParameters<Transport, TChain, Account, TAbi>["abi"];
```
hmm without this abi typing, getContract() returned contract object does not have method functions typed, so useful to have it here

denniswon commented:
All the comments are accommodated, and this PR should have all necessary implementations for both 0.6 and 0.7 entry point support with strong typing of user operations, etc., with default entry point support for version (0.6) and address.

Next step:
Thanks for the first-round review; I would appreciate reviewing the overall design. Please note that there are follow-up PRs for making typing less strict in some places with fallbacks.
- Unit test PR is included in the follow-up PR.
- With the default 0.6 entry point, all the existing unit/e2e tests should pass for migration safety.
- Documentation update for different entry point support

cc: @moldy530 @dancoombs @dphilipson @rthomare

moldy530 commented on packages/accounts/src/light-account/account.ts:
```diff
@@ -32,45 +37,72 @@ import {
 };
 
 export type CreateLightAccountParams<
+  TEntryPointVersion extends EntryPointVersion,
```
I don't think we should have added these new generics as the first parameter in these types. These are technically breaking changes

moldy530 commented on packages/accounts/src/light-account/account.ts:
```diff
@@ -32,45 +37,72 @@ import {
-  ToSmartContractAccountParams<"LightAccount", TTransport>,
-  "transport" | "chain" | "entryPoint" | "accountAddress"
+  ToSmartContractAccountParams<TEntryPointVersion, "LightAccount", TTransport>,
+  "transport" | "chain"
```
wait, why remove the `accountAddress` from this pick?

moldy530 replied:
just checked it locally and I'm not seeing an error by keeping it

moldy530 replied:
why was this resolved? mind adding reasons when resolving comments? it's helpful to have more context. if there's an error it's helpful to include a screenshot of the error or console dump of the error

moldy530 replied:
bump

denniswon replied:
ah missed. was actually done on the child PR, but as I refactored fixes into here, that reverted .

moldy530 commented on packages/accounts/src/light-account/account.ts:
```diff
@@ -32,45 +37,72 @@ import {
 } from "./utils.js";
 
 export type LightAccount<
+  TEntryPointVersion extends EntryPointVersion,
```
LightAccount should default to 0.6. Actually, it shouldn't just default but require it for now because LightAccount doesn't support 0.7

moldy530 commented on packages/accounts/src/light-account/account.ts:
```diff
@@ -32,45 +37,72 @@ import {
   initCode,
   version = "v1.1.0",
-  entryPoint = getVersion060EntryPoint(chain),
+  entryPoint,
```
yea we should just default to 0.6 here and remove the generics for entrypoint def for light account and MA as well. they all require 0.6

moldy530 commented on packages/accounts/src/light-account/account.ts:
```diff
@@ -160,74 +192,148 @@ export async function createLightAccount({
-          accum[0].push(curr.target);
-          accum[1].push(curr.value ?? 0n);
-          accum[2].push(curr.data);
+  const account =
```
there's a lot of duplicated code here. We should leave `createLightAccount` as it was and have it just require 0.6 for now. in a later PR we should tackle supporting LightAccount v2 and do it in such a way that we're not duplicating a lot of code

dphilipson replied:
I agree about the duplicated code, but the contracts team is also hoping that they can use the SDK to test light account v2 so they would like support here even before it's deployed. I'll get some more info about what they're looking for.

moldy530 replied:
makes sense, we should do the 0.7 support for LA as a separate PR to this so that it's easier to parse and review

moldy530 commented on packages/accounts/src/light-account/actions/transferOwnership.ts:
```diff
@@ -46,13 +58,14 @@ export const transferOwnership: <
-      account,
     },
     account,
+    overrides,
```
good catch

moldy530 commented on packages/accounts/src/light-account/client.ts:
```diff
@@ -18,49 +19,81 @@ import {
 } from "./decorator.js";
 
 export type CreateLightAccountClientParams<
-  TTransport extends Transport = Transport,
```
wait we still need this here

moldy530 commented on packages/accounts/src/light-account/client.ts:
```diff
@@ -18,49 +19,81 @@ import {
   transport,
   chain,
   ...clientConfig
-}: CreateLightAccountClientParams): Promise<SmartAccountClient> {
```
we should leave this as it was before without specifying generics, the point of the overload was so that we wouldn't get typing issues when using combinations of chains, signers, or transports

moldy530 commented on packages/accounts/src/light-account/client.ts:
```diff
@@ -18,49 +19,81 @@ import {
 >;
 
-export async function createLightAccountClient({
+export async function createLightAccountClient<
```
same here, leave this as it was

moldy530 reviewed (commented):
Gonna do another review, but meta comment on this initial pass is that:
1. we should be adding `TEntrypoint` generics at the end of the list so that the change isn't a breaking change
2. All of the existing accounts should be requiring 0.6 for now which means most of them don't even need the generic

moldy530 commented on packages/alchemy/src/client/internal/smartAccountClientFromRpc.ts:
```diff
@@ -45,7 +49,15 @@ export function createAlchemySmartAccountClientFromRpcClient({
   gasEstimator,
   customMiddleware,
   client,
-}: CreateAlchemySmartAccountClientFromRpcClient): AlchemySmartAccountClient {
```
this override as well should use the default generic values and not also be generic on these types

moldy530 commented on packages/alchemy/src/client/internal/smartAccountClientFromRpc.ts:
```diff
@@ -79,5 +91,10 @@ export function createAlchemySmartAccountClientFromRpcClient({
         },
       })),
-  }).extend(alchemyActions);
+  }).extend(alchemyActions) as AlchemySmartAccountClient<
```
if you leave them as defaults you don't have to do the typecast here

moldy530 commented on packages/accounts/src/msca/account/multiOwnerAccount.ts:
```diff
@@ -90,27 +131,44 @@ export async function createMultiOwnerModularAccount({
-    ...multiOwnerMessageSigner(client, accountAddress, () => signer),
   });
 
+  const baseAccount =
```
same comment here as with light account, this doesn't support 0.7 so we shouldn't even allow devs to do this

moldy530 commented on packages/core/src/account/smartContractAccount.ts:
```diff
@@ -40,10 +47,11 @@ export type AccountOp = {
 export type GetAccountParameter<
-  TAccount extends SmartContractAccount | undefined =
-    | SmartContractAccount
+  TEntryPointVersion extends EntryPointVersion,
```
I think we should set a default value here like we do everywhere. something like `TEntryPointVersion extends EntryPointVersion = EntrypointVersion` and do that everywhere. this lets us add it to the end of the list to be backwards compatible and makes it a bit easier to work with.

then specific impls can also restrict the version by using 0.6 or 0.7 or keeping the default to signal they support both

moldy530 replied:
oh for this param in particular too, if it's a default on `SmartContractAccount` then `GetAccountParameter` doesn't even need this generic

moldy530 commented on packages/core/src/account/smartContractAccount.ts:
```diff
@@ -54,16 +62,17 @@ export type UpgradeToAndCallParams = {
 export type SmartContractAccount<
-  Name extends string = string,
-  TUO = UserOperationRequest
+  TEntryPointVersion extends EntryPointVersion,
```
see comment above, meant to add it here

moldy530 commented on packages/core/src/account/smartContractAccount.ts:
```diff
@@ -82,21 +91,28 @@ export type SmartContractAccount<
 };
 //#endregion SmartContractAccount
 
+export interface AccountEntryPointRegistry<Name extends string = string>
```
how is this used?

moldy530 replied:
looking through the code base this isn't used anywhere actually, we should drop this

moldy530 replied:
why was this resolved? This isn't used anywhere so why keep it? This type is already supported by just having `SmartContractAccount<Name, TEntryPointVersion>` and is how we do this in most places

denniswon replied:
this is removed.

moldy530 commented on packages/core/src/account/smartContractAccount.ts:
```diff
@@ -314,30 +377,58 @@ export async function toSmartContractAccount<
+    throw new InvalidEntryPointError(chain, _entryPoint.version);
+  }
+
+  return _entryPoint.version === "0.6.0"
```
a lot of duplicated code here too. why do we need the typecast?

moldy530 commented on packages/core/src/account/smartContractAccount.ts:
```diff
@@ -179,24 +202,61 @@ export async function toSmartContractAccount<
+  TTransport,
+  TChain
+>): Promise<AccountEntryPointRegistry<Name>[EntryPointVersion]> {
+  const _entryPoint: EntryPointDef<EntryPointVersion> =
+    entryPoint ?? getEntryPoint(chain);
```
I think we should make `entryPoint` a required parameter instead so that typescript infers the output

dphilipson commented on packages/accounts/src/msca/account/multiOwnerAccount.ts:
```diff
@@ -90,27 +131,44 @@ export async function createMultiOwnerModularAccount({
+          getAccountInitCode,
+          ...standardExecutor,
+          ...multiOwnerMessageSigner(client, _accountAddress, () => signer),
+        })
```
We should look for ways to reduce repetition between the branches. One way to do that might be to factor out all the common props:
```ts
  const commonParams: Omit<
    Parameters<typeof toSmartContractAccount>[0],
    "entryPoint"
  > = {
    transport,
    chain,
    accountAddress: _accountAddress,
    source: `MultiOwnerModularAccount`,
    getAccountInitCode,
    ...standardExecutor,
    ...multiOwnerMessageSigner(client, _accountAddress, () => signer),
  };

  const baseAccount =
    _entryPoint.version === "0.6.0"
      ? await toSmartContractAccount({
          ...commonParams,
          entryPoint: _entryPoint as EntryPointDef<"0.6.0">,
        })
      : _entryPoint.version === "0.7.0"
      ? await toSmartContractAccount({
          ...commonParams,
          entryPoint: _entryPoint as EntryPointDef<"0.6.0">,
        })
      : undefined;
```
(But as Moldy said, this particular account should be 0.6.0 only.)

dphilipson replied:
Actually wait a sec, isn't this just the same value in both cases? Can you do
```ts
  const baseAccount = await toSmartContractAccount({
    transport,
    chain,
    entryPoint: _entryPoint as EntryPointDef<"0.6.0"> | EntryPointDef<"0.7.0">,
    accountAddress: _accountAddress,
    source: `MultiOwnerModularAccount`,
    getAccountInitCode,
    ...standardExecutor,
    ...multiOwnerMessageSigner(client, _accountAddress, () => signer),
  });
```

moldy530 replied:
modular account doesn't support 0.7 either does it? so we shouldn't be doing this here. 

the default entrypoint version type in https://github.com/alchemyplatform/aa-sdk/pull/541 is the union which means that if the code is the same for both then you wouldn't even need the typecast and you'd be able to leave this file as is without any changes

moldy530 commented on packages/core/src/types.ts:
```diff
@@ -241,4 +322,54 @@ export interface UserOperationStruct {
+    : never;
 //#endregion UserOperationStruct
+
+export type UserOperationLike<
```
this type is kinda confusing because it's checking only against UserOperationOverrides, I expected it to be more like UserOperationStruct or Request.

How it used and why is it used over the overrides?

denniswon replied:
using overrides should work technically, but I want to check if all fields exist, even if undefined value fields.

moldy530 replied:
what I'm saying it only works with Overrides, but not the other UserOperation type structs. I think we should update the name here

denniswon replied:
do not really need this anymore now. so this is removed. Think this causes more confusion, so better to be clearer.

moldy530 commented on packages/core/src/entrypoint/index.ts:
```diff
@@ -0,0 +1,190 @@
+      isUserOpVersion: v6.isUserOpVersion,
+      coerce: (userOperation, throwable) =>
+        coerceEntryPointVersion(v6, userOperation, throwable),
+    } as EntryPointDef<"0.6.0", TChain, typeof EntryPointAbi_v6>;
```
better we use `satisfies` here instead of `as` for some additional type safety

denniswon replied:
can't use satisfies here. typescript errors.

moldy530 replied:
see below, we should fix the typescript errors that keeps satisfies from working

moldy530 replied:
see https://github.com/alchemyplatform/aa-sdk/pull/549

the reason satisfies was not working is because the resulting object doesn't contain an `overrides` property which is required (though can be set to undefined explicitly) we should fix that.

moldy530 commented on packages/core/src/entrypoint/index.ts:
```diff
@@ -0,0 +1,190 @@
+      isUserOpVersion: v7.isUserOpVersion,
+      coerce: (userOperation, throwable) =>
+        coerceEntryPointVersion(v7, userOperation, throwable),
+    } as EntryPointDef<"0.7.0", TChain, typeof EntryPointAbi_v7>;
```
same here should use `satisfies` instead of `as`

denniswon replied:
no you need to use as. satisfies does work here.

moldy530 replied:
if satisfies isn't working it's because the type doesn't match the expected output... we should fix that so it does work

moldy530 replied:
see https://github.com/alchemyplatform/aa-sdk/pull/549

moldy530 commented on packages/core/src/entrypoint/index.ts:
```diff
@@ -0,0 +1,190 @@
+    throw new EntryPointNotFoundError(chain, version);
+  }
+
+  if (version === "0.6.0") {
+    const v6 = entryPoint as SupportedEntryPoint<
+      "0.6.0",
+      TChain,
+      typeof EntryPointAbi_v6
+    >;
```
you can avoid the typecast and need for the `v6` const completely here by just checking:
```
if (entryPoint.version === "0.6.0")
```

This will then allow typescript to infer the the type of `entryPoint` is already the v6 one and same below

denniswon replied:
no, this one is not a entrypoint version generic function. just checking for version === '0.6.0' does not return the type inferred result.

moldy530 replied:
sorry I'm not saying use `version === 0.6.0` but `entryPoint.version === 0.6.0` that will be inferred correctly

moldy530 replied:
not only does it work, but if you do that you don't even need the typecasting. 

Typecasting is generally a code smell. if things are typed correctly and the logic adheres by to the types, then you shouldn't need to do any of that.

because it's easier than making the suggestions in github this is what I mean: https://github.com/alchemyplatform/aa-sdk/pull/549

moldy530 commented on packages/accounts/plugingen/phases/plugin-actions/read-actions.ts:
doesn't look like this and the cli tool's file for `read-actions` are in sync anymore

moldy530 commented on packages/plugingen/src/commands/generate/phases/plugin-actions/index.ts:
```diff
@@ -46,49 +46,53 @@ export const PluginActionsGenPhase: Phase = async (input) => {
+          ? dedent`{
+                  args,
+                  overrides,
+                  account = client.account
```
looks like `context` is missing here

denniswon replied:
[I don't think I need it because it's not used in ](https://github.com/alchemyplatform/aa-sdk/pull/514/files#r1562043130)

moldy530 replied:
seems like a bug then because the original code does include it: https://github.com/alchemyplatform/aa-sdk/pull/514/files#diff-67ce60c0c18c8684ea92aadf41e3e203bb0b198f3bb0615b56fe91cb72ce612eR53

moldy530 replied:
can you add the `context` parameter to both of these lines? it should match the code in the other `plugin-actions/index` file

moldy530 commented on packages/core/src/actions/smartAccount/dropAndReplaceUserOperation.ts:
```diff
@@ -54,23 +75,33 @@ export const dropAndReplaceUserOperation: <
       BigInt(maxFeePerGas ?? 0n),
-      bigIntMultiply(uoToDrop.maxFeePerGas, 1.1)
+      bigIntMultiply(
+        (uoToDrop as UserOperationRequest<TEntryPointVersion>).maxFeePerGas,
```
do we need to type cast here? can we not just do `uoToDrop.maxFeePerGas`?

moldy530 replied:
confirmed locally this typescast is not needed

moldy530 commented on packages/core/src/actions/smartAccount/dropAndReplaceUserOperation.ts:
```diff
@@ -54,23 +75,33 @@ export const dropAndReplaceUserOperation: <
       BigInt(maxPriorityFeePerGas ?? 0n),
-      bigIntMultiply(uoToDrop.maxPriorityFeePerGas, 1.1)
+      bigIntMultiply(
+        (uoToDrop as UserOperationRequest<TEntryPointVersion>)
+          .maxPriorityFeePerGas,
```
same here

moldy530 replied:
confirmed locally this typecast is not needed

moldy530 commented on packages/core/src/actions/smartAccount/internal/sendUserOperation.ts:
```diff
@@ -30,16 +34,27 @@ export const _sendUserOperation: <
   }
 
+  request.signature = await account.signUserOperationHash(
+    entryPoint.version === "0.6.0"
+      ? entryPoint.getUserOperationHash(
+          request as UserOperationRequest<"0.6.0">
+        )
+      : entryPoint.getUserOperationHash(
+          request as UserOperationRequest<"0.7.0">
+        )
```
just tested this locally and you should be able to just do:
```suggestion
    entryPoint.getUserOperationHash(request)
```

moldy530 replied:
bump

moldy530 commented on packages/core/src/actions/smartAccount/sendTransaction.ts:
```diff
@@ -55,8 +62,8 @@ export const sendTransaction: <
   const { hash } = await _sendUserOperation(client, {
     account: account as SmartContractAccount,
-    uoStruct,
+    uoStruct: uoStruct as UserOperationStruct<TEntryPointVersion>,
```
tested locally and you don't need this typecast here:
```suggestion
    uoStruct,
```

denniswon replied:
updated

moldy530 replied:
bump, still seeing this here?

moldy530 commented on packages/core/src/entrypoint/0.6.ts:
```diff
@@ -1,28 +1,105 @@
+  },
+
+  packUserOperation,
+} as SupportedEntryPoint<"0.6.0", Chain, typeof EntryPointAbi_v6>;
```
```suggestion
} satisfies SupportedEntryPoint<"0.6.0", Chain, typeof EntryPointAbi_v6>;
```

denniswon replied:
udpated

moldy530 commented on packages/core/src/entrypoint/0.7.ts:
```diff
@@ -0,0 +1,172 @@
+  },
+
+  packUserOperation,
+} as SupportedEntryPoint<"0.7.0", Chain, typeof EntryPointAbi_v7>;
```
```suggestion
} satisfies SupportedEntryPoint<"0.7.0", Chain, typeof EntryPointAbi_v7>;
```

denniswon replied:
updated

moldy530 replied:
bump, still seeing this as `as`

moldy530 commented on packages/core/src/utils/types.ts:
```diff
@@ -37,3 +38,57 @@ export type Prettify<T> = {
+export type OneOf<T1, T2> = IsOneOf<T1, T2> extends true ? T1 : never;
+
+export type Interface<T> = {
+  [K in keyof T]: any;
```
should this actually be:
```suggestion
  [K in keyof T]: T[K];
```

denniswon replied:
this type is removed as unused anymore. UserOperationLike is removed.

moldy530 replied:
bump, I still see it here?

denniswon commented on packages/accounts/plugingen/phases/plugin-actions/index.ts:
```diff
@@ -43,15 +47,15 @@ export const PluginActionsGenPhase: Phase = async (input) => {
+          ? dedent`{
+                  args,
+                  overrides,
                   context,
                   account = client.account
               }`
-          : dedent`{ 
-                  overrides, 
-                  account = client.account 
+          : dedent`{
+                  overrides,
+                  account = client.account
```
@moldy530 context is not needed here, wasn't here before either.

moldy530 replied:
it's a bug if it wasn't... see the above part of the ternary. that one includes `context`

denniswon replied:
when there is no input, do we still need context?

denniswon replied:
ah I see. I guess we dont have any plugin action methods with no input arg? Added context var here.

denniswon commented on packages/accounts/src/msca/plugin-manager/installPlugin.ts:
```diff
@@ -70,7 +71,12 @@ export async function installPlugin<
+  return client.sendUserOperation({
+    uo: callData,
+    overrides,
+    account,
```
added

denniswon commented on packages/accounts/src/msca/plugin-manager/uninstallPlugin.ts:
```diff
@@ -61,7 +64,12 @@ export async function uninstallPlugin<
+  return client.sendUserOperation({
+    uo: callData,
+    overrides: overrides,
+    account,
```
added

denniswon commented on packages/alchemy/src/client/smartAccountClient.ts:
```diff
@@ -55,14 +56,15 @@ export type AlchemySmartAccountClient_Base<
     | Record<string, any>
-    | undefined
+    | undefined,
+  version extends GetEntryPointFromAccount<account> = GetEntryPointFromAccount<account>
```
updated :) good point

denniswon commented on packages/alchemy/src/client/types.ts:
```diff
@@ -9,14 +10,14 @@ import type {
         policyId: string;
         entryPoint: Address;
-        userOperation: UserOperationRequest;
+        userOperation: UserOperationRequest<TEntryPointVersion>;
```
updated :)

denniswon commented:
@moldy530 per your comment,

> https://github.com/alchemyplatform/aa-sdk/pull/514#discussion_r1558157699

SupportedEntryPoint contains a map of chained => EP address. But this is different from EntryPointDef.

getEntryPoint() first fetch the supported entry point for the version, then with the input chain param, gets the EP address from the above EP address map unless the user inputs addressOverride.

So the user can do `getEntryPoint(chain, { addressOverride: '0xasdfasdfasdf' })` for whatever chain they use as long as the bundler/infra supports.

denniswon commented:
@moldy530 I refactored the light account v2 pr, moved all entrypoint v7 base branch related fixes/changes on this PR on the HEAD commit. tested with unit test/e2e test, etc. per our sync today, for v6 backward compatibility, nothing breaking. 

Ready for final pass, and merge to main branch.

cc: @dphilipson @dancoombs @dphilipson @rthomare

denniswon commented:
Lint and Test passes, but PR title format is not passing even after changing the PR title to fix the fail issue. Should be okay. @moldy530

moldy530 commented on packages/alchemy/src/client/types.ts:
```diff
@@ -49,43 +72,68 @@ export type AlchemyRpcSchema = [
+          | "maxPriorityFeePerGas"
+        > &
+          (
+            | {
+                paymasterAndData: UserOperationRequest<"0.6.0">["paymasterAndData"];
+              }
+            | Pick<
+                UserOperationRequest<"0.7.0">,
+                | "paymaster"
+                | "paymasterData"
+                | "paymasterPostOpGasLimit"
+                | "paymasterVerificationGasLimit"
+              >
+          )
```
nit: instead of copy pasting this type, can we wrap this in a different type definition so it can be reused?

moldy530 commented on packages/alchemy/src/config/actions/createAccount.ts:
```diff
@@ -64,6 +65,7 @@ export async function createAccount<TAccount extends SupportedAccountTypes>(
           signer,
           transport: (opts) => transport({ ...opts, retryCount: 0 }),
           chain,
+          entryPoint: getEntryPoint(chain),
```
why is this necessary? this is already part of the `params` object

moldy530 commented on packages/alchemy/src/middleware/userOperationSimulator.ts:
```diff
@@ -2,21 +2,22 @@ import {
       method: "alchemy_simulateUserOperationAssetChanges",
       params: [
-        deepHexlify(await resolveProperties(struct)) as UserOperationStruct,
+        deepHexlify(
+          await resolveProperties(struct)
+        ) as UserOperationStruct<TEntryPointVersion>,
```
doesn't `alchemy_simulateUserOpeartionAssetChanges` support a union type? if so we should be able to leave this as it was and not pass in the generic here

moldy530 replied:
if not, then we should instead of doing `TEntryPointVersion extends EntryPointVersion` do `extends GetEntryPointFromAccount`

denniswon replied:
ah its not generic anymore. will remove

moldy530 commented on packages/core/e2e-tests/constants.ts:
@dancoombs made a comment about this in slack. Rather we change to using an `API_URL` env var than adding staging env vars so that this is easier to do with just one env var

moldy530 commented on packages/core/src/__tests__/utils.test.ts:
```diff
@@ -1,38 +1,31 @@
+      } as UserOperationRequest<"0.6.0">)
     ).toMatchInlineSnapshot(
-      `"0xa70d0af2ebb03a44dcd0714a8724f622e3ab876d0aa312f0ee04823285d6fb1b"`
+      '"0xbb5560c1a3983429a6cdb244fa532fb4f2cf0de8ba9ccbf257bff93d069c76a3"'
```
bump

moldy530 commented on packages/core/src/account/base.ts:
```diff
@@ -379,7 +380,7 @@ export abstract class BaseSmartContractAccount<
     }
 
-    return trim(storage);
+    return toHex(trim(storage));
```
this is a bug I think. I ran into this issue somewhere else. This needs to just be `trim` and not `toHexed` again. The issue I ran into was that if you `toHex` the hex string it generates a new `hex` which is the hex representation of the bytes of the previous hex string

moldy530 commented on packages/core/src/account/types.ts:
```diff
@@ -135,6 +135,7 @@ export interface ISmartContractAccount<
   getSigner(): TSigner;
 
   /**
+   *
```
nit: either remove this or add a description here

moldy530 commented on packages/core/src/actions/bundler/estimateUserOperationGas.ts:
```diff
@@ -1,16 +1,18 @@
 } from "../../types";
 
 export const estimateUserOperationGas = async <
+  TEntryPointVersion extends EntryPointVersion,
```
nit: for consistency with the alchemy simulate UO method and the other methods, I think this should go second in the list. Also this is for the same reason as my previous comments on the original PR -- if anyone was explicitly typing this method for w/e reason, this is a breaking change for them

moldy530 commented on packages/core/src/actions/bundler/sendRawUserOperation.ts:
```diff
@@ -1,13 +1,15 @@
 import type { UserOperationRequest } from "../../types";
 
 export const sendRawUserOperation = async <
+  TEntryPointVersion extends EntryPointVersion,
```
same comment as above

moldy530 commented on packages/core/src/actions/smartAccount/internal/runMiddlewareStack.ts:
```diff
@@ -22,38 +24,52 @@ const asyncPipe =
 
-  return resolveProperties<UserOperationStruct>(result);
-};
+  return resolveProperties(result) as Promise<
+    UserOperationStruct<GetEntryPointFromAccount<TAccount>>
+  >;
```
typecasting should be avoided when it is possible
```suggestion
  return resolveProperties<
    UserOperationStruct<GetEntryPointFromAccount<TAccount>>
  >(result);
```

moldy530 commented on packages/core/src/actions/smartAccount/internal/sendUserOperation.ts:
```diff
@@ -30,16 +34,29 @@ export const _sendUserOperation: <
     throw new InvalidUserOperationError(args.uoStruct);
   }
 
+  console.log("request", client.chain.name, request);
```
remove

denniswon replied:
removed from the child pr. sorry

moldy530 commented on packages/core/src/client/decorators/bundlerClient.ts:
```diff
@@ -58,8 +68,10 @@ export type BundlerActions = {
-  estimateUserOperationGas(
-    request: UserOperationRequest,
+  estimateUserOperationGas<
+    TEntryPointVersion extends EntryPointVersion = DefaultEntryPointVersion
```
```suggestion
    TEntryPointVersion extends EntryPointVersion = EntryPointVersion
```

moldy530 replied:
this one I think we talked about async, but this method should probs not default

moldy530 commented on packages/core/src/client/decorators/bundlerClient.ts:
```diff
@@ -71,8 +83,10 @@ export type BundlerActions = {
-  sendRawUserOperation(
-    request: UserOperationRequest,
+  sendRawUserOperation<
+    TEntryPointVersion extends EntryPointVersion = DefaultEntryPointVersion
```
this one I think we talked about async, but this method should probs not default
```suggestion
    TEntryPointVersion extends EntryPointVersion = EntryPointVersion
```

moldy530 commented on packages/core/src/entrypoint/types.ts:
```diff
@@ -1,9 +1,124 @@
+  : {
+      addressOverride?: Address;
+      version: OneOf<TEntryPointVersion, EntryPointVersion>;
+    };
```
bump

moldy530 commented on packages/core/src/index.ts:
```diff
@@ -92,7 +94,7 @@ export {
   InvalidRpcUrlError,
 } from "./errors/client.js";
-export { EntryPointNotFoundError as EntrypointNotFoundError } from "./errors/entrypoint.js";
+export * from "./errors/entrypoint.js";
```
we should not be `export *` anything that isn't a type

moldy530 commented on packages/core/src/index.ts:
```diff
@@ -71,8 +73,8 @@ export {
 } from "./ens/utils.js";
-export { getVersion060EntryPoint } from "./entrypoint/0.6.js";
-export { type EntryPointDef } from "./entrypoint/types.js";
+export * from "./entrypoint/index.js";
```
we should not be `export *` anything that isn't a type

moldy530 commented on packages/core/src/middleware/defaults/feeEstimator.ts:
```diff
@@ -18,7 +18,7 @@ export const defaultFeeEstimator: <C extends MiddlewareClient>(
 
     const feeData = await client.estimateFeesPerGas();
-    if (!feeData.maxFeePerGas || !feeData.maxPriorityFeePerGas) {
+    if (!feeData.maxFeePerGas || feeData.maxPriorityFeePerGas == null) {
```
think I left a comment about this somewhere else, but nit:

```suggestion
    if (feeData.maxFeePerGas == null || feeData.maxPriorityFeePerGas == null) {
```

moldy530 replied:
we should be consistent

denniswon replied:
hmm this was because maxFeePerGas cannot be 0, but maxPriorityFeePerGas can be

moldy530 replied:
yea that's fine, if one can be 0 and not the other checking both explicitly for `== null` still covers the one that can't be 0 and at least they read the same.

it's nit, so it's fine either way

moldy530 commented on packages/core/src/middleware/defaults/overridePaymasterData.ts:
```diff
@@ -1,10 +1,34 @@
+          : (struct as UserOperationStruct<"0.6.0">).paymasterAndData;
+      break;
+    case "0.7.0":
+      const _overrides = overrides as UserOperationOverrides<"0.7.0">;
```
why not do the same above? it's much easier to read

moldy530 commented on packages/core/src/types.ts:
```diff
@@ -52,34 +54,70 @@ export type UserOperationFeeOptions = z.input<
+    nonceKey: bigint;
+    stateOverride: StateOverride;
+  } & (TEntryPointVersion extends "0.6.0"
+    ? {
+        paymasterAndData: UserOperationStruct<"0.6.0">["paymasterAndData"];
+      }
+    : TEntryPointVersion extends "0.7.0"
+    ?
+        | {
+            paymaster: NoUndefined<UserOperationStruct<"0.7.0">["paymaster"]>;
+            paymasterData: NoUndefined<
+              UserOperationStruct<"0.7.0">["paymasterData"]
+            >;
+            paymasterPostOpGasLimit: NoUndefined<
+              UserOperationStruct<"0.7.0">["paymasterPostOpGasLimit"]
+            >;
+            paymasterVerificationGasLimit: NoUndefined<
+              UserOperationStruct<"0.7.0">["paymasterVerificationGasLimit"]
+            >;
+          }
+        | undefined
```
yea we do this in A LOT of places, can we make this a type alias (think I left a comment above about this too)

denniswon commented:
@moldy530 

> https://github.com/alchemyplatform/aa-sdk/pull/514#discussion_r1562739793

sorry this is removed from the light account v2 pr

moldy530 commented on packages/alchemy/src/client/types.ts:
```diff
@@ -16,30 +21,52 @@ export type AlchemyRpcSchema = [
     ];
-    ReturnType: { paymasterAndData: Hex };
+    ReturnType:
+      | {
+          paymasterAndData: UserOperationRequest<"0.6.0">["paymasterAndData"];
+        }
+      | Pick<
+          UserOperationRequest<"0.7.0">,
+          | "paymaster"
+          | "paymasterData"
+          | "paymasterPostOpGasLimit"
+          | "paymasterVerificationGasLimit"
+        >;
```
left a comment about this somewhere in the last review, but let's make this a type alias and use it here and below and anywhere else we need a union type for the `PaymasterAndDataFields`

moldy530 commented on packages/core/src/account/types.ts:
```diff
@@ -136,11 +136,16 @@ export interface ISmartContractAccount<
 
   /**
    * @returns the address of the factory contract for the smart account
+   *
+   * The factory address of the smart account created by the factory
```
sorry for this nit... can you move this above the `@returns`? This will be a linting error once I get my jsdocs linter in

Order should be:
- description
- param(s)
- returns

moldy530 commented on packages/core/src/account/types.ts:
```diff
@@ -136,11 +136,16 @@ export interface ISmartContractAccount<
 
   /**
    * @returns the address of the entry point contract for the smart account
+   *
+   * The entrypoint address is the address ERC 4337 entrypoint singleton contract
+   * that this smart account and the connected bundler client use
```
same here

moldy530 reviewed (commented):
last stretch!

denniswon commented:
@moldy530 accommodated all the last stretch items.
